### DISCUSSION TITLE
Hono RPC Capabilities

### DISCUSSION THREAD

**Discussion by Rykuno** [Savory/Danet#95] [Ideas]
Of my three recent projects I've used Hono RPC and its been a complete lifesaver and godsend for e2e typesafety. Is there any possibility to add its capability to the Danet framework? 

I get that openAPI specs exist and generators can be used to generate types/fetch calls - but from my experience they have issues nearly everytime a project's types get relatively complex.

**Comment by Sorikairox** [comment]
@Rykuno Interesting. I've never used RPC before, but I know what it does. Do you have any insight on how integrating it to Danet could look like ?

**Reply by Rykuno** [reply]
I do! Its very easy but there is a prerequisite that essentially involves either registering routes/paths to a centralized hono instance or chaining routes together and returning that instance's type to be used by `@hono/client` in some other application. 

The Hono RPC type can either be exported in a monorepo or published as a type to a private repository if not for absolute full e2e typesafety. 

I've generated types from openAPI specs for years and I can't even begin to express how much more helpful this is if your full stack is in deno/node.

Here is an example of how I do it(in a knockoff nestjs fashion). 

This defines the type from the hono routes in my api,
https://github.com/Rykuno/TofuStack/blob/main/src/lib/server/api/index.ts

Then in my client, I simply register it as the following,
```
import { hc } from 'hono/client';
export apiClient =  hc<ApiRoutes>('/').api;
```
---

I haven't looked at how Danet register's routes via the declarations but I'm guessing chaining routes and returning a single instance of hono might be difficult. If so, factories could be a much more practical answer that allows a bit more decentralized solution. 

Essentially each HTTP declaration would registered as a factory handler stored and globally accessed somewhere.
https://hono.dev/docs/helpers/factory#factory-createhandlers

This is all purely speculative atm because i havent looked at the Danet source, but I'll check it out tomorrow!

**Reply by Sorikairox** [reply]
@Rykuno I see !

Given that Danet expose the underlying hono app via `yourDanetApplication.router()`, it might be easy ?
